core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -1581,6 +1650,178 @@ private void validateTableModelAndCreateTypeCast(
+
+                tok = SqlUtil.fetchNext(lexer);
+                if (null == tok || Chars.equals(tok, ';')) {
+                    break;
+                }
+                if (!Chars.equals(tok, ',')) {
+                    throw SqlException.position(lexer.lastTokenPosition()).put(" expected ','");
+                }
+            }
+
+            for (int n = 0; n < tableNames.size(); n++) {
+                backupTable(tableNames.get(n), executionContext);
+            }
+
+            return compiledQuery.ofBackupTable();
+        } finally {
+            tableNames.clear();
+        }
+    }
+
+    private void setupBackupRenamePath() {
+        TimestampFormat format = configuration.getBackupDirTimestampFormat();
+        long epochMicros = configuration.getMicrosecondClock().getTicks();
+        int n = 0;
+        // TODO: There is a race here, two threads could try and create the same renamePath, only one will succeed the other will throw
+        // a CairoException. Maybe it should be serialised
+        do {
+            renamePath.of(configuration.getBackupRoot()).put(Files.SEPARATOR);
+            format.format(epochMicros, TimestampLocaleFactory.INSTANCE.getDefaultTimestampLocale(), null, renamePath);
+            if (n > 0) {
+                renamePath.put('.').put(n);
+            }
+            renamePath.put(Files.SEPARATOR).$();
+            n++;
+        } while (ff.exists(renamePath));
+        if (ff.mkdirs(renamePath, configuration.getMkDirMode()) != 0) {
+            throw CairoException.instance(ff.errno()).put("could not create [dir=").put(renamePath).put(']');
+        }
+    }
+
+    private transient String cachedTmpBackupRoot;
+
+    private void backupTable(@NotNull CharSequence tableName, @NotNull SqlExecutionContext executionContext) {
+        LOG.info().$("Starting backup of ").$(tableName).$();
+        if (null == cachedTmpBackupRoot) {
+            if (null == configuration.getBackupRoot()) {
+                throw CairoException.instance(0).put("Backup is disabled, no backup root directory is configured in the server configuration ['cairo.sql.backup.root' property]");
+            }
+            path.of(configuration.getBackupRoot()).concat(configuration.getBackupTempDirName()).put(Files.SEPARATOR).$();
+            cachedTmpBackupRoot = path.toString();
+            if (!ff.exists(path)) {
+                LOG.info().$("Creating backup directory ").$(cachedTmpBackupRoot).$();
+                if (ff.mkdirs(path, configuration.getMkDirMode()) != 0) {
+                    throw CairoException.instance(ff.errno()).put("Could not create [dir=").put(path).put(']');
+                }
+            }
+        }
+
+        CairoSecurityContext securityContext = executionContext.getCairoSecurityContext();
+        try (TableReader reader = engine.getReader(securityContext, tableName)) {
+            cloneMetaData(tableName, configuration.getRoot(), cachedTmpBackupRoot, configuration.getMkDirMode(), reader);
```

Comment:
on the subject of mkdir mode, perhaps we should have separate configuration property for backup mode?

Original mkdirMode is for directories created by server runtime, these are not necessarily need to be readable by other processes and/or users. On other hand backup directory could potentially be picked up by another process, such as tape backup or something similar and thus need to have different permissions than runtime dirs.